pkg/server/endpoints/eventTracker.go

```diff
@@ -0,0 +1,257 @@
+package endpoints
+
+import (
+	"maps"
+	"slices"
+	"time"
+)
+
+/**
+ * Tracks events as they indivicually walk through a list of event boundaries.
+ *
+ * An event track is defined with a set of boundaries, which are indexes to
+ * virtual hash tables, with the event's hash determining the position within
+ * that hash table where the event will be selected to be polled.
+ * For eventTRackers that lack boundaries, or polls that exist prior to
+ * boundaries, the event is always polled.
+ */
+type eventTracker struct {
+	/* Times the event is polled before entering a boundary */
+	initialPolls uint
+	/* Times the event is polled */
+	pollPeriods uint
+	/* Per event context of each event's walk across the boudaries */
+	events map[uint]*eventStats
+	/* The leading index of boundaries in which an event should only report once */
+	boundaries []uint
+}
+
+/**
+ * Tracks event context in its walk of the event boundaries.
+ */
+type eventStats struct {
+	/* The event's hash for hash table calcuations */
+	hash uint
+	/* The number of times the event was considered for polling */
+	ticks uint
+	/* The number of times the event was selected for polling */
+	polls uint
+}
+
+/**
+ * A utility function to get the number of PollTimes (ticks) in an interval.
+ *
+ * Subsecond inputs are adjusted to a minimum value one second.
+ *
+ * @returns One tick, or the smallest number of ticks that just exceeds the trackTime..
+ */
+func PollPeriods(pollTime time.Duration, trackTime time.Duration) uint {
+	if pollTime < (time.Duration(1) * time.Second) {
+		pollTime = time.Duration(1) * time.Second
+	}
+	if trackTime < (time.Duration(1) * time.Second) {
+		trackTime = time.Duration(1) * time.Second
+	}
+	return uint(1 + (trackTime-1)/pollTime)
+}
+
+/**
+ * The default boundary strategy.
+ *
```

Comment:
This strategy is completely arbitrary, in the sense I created it from nothing, using only a vauge idea of what might be required, because the people asking for the backoff also didn't have any concrete ideas of what they wanted either.

That's because an exponential backoff won't work (it was initially suggested).  With exponential, your cache can drift from the database state far beyond what operations would permit (30 minutes to 8 hours to detect a late arriving change is quite possibly with exponential).

So, I figured, 1 minute a the configured polling, 9 minutes of twice per minute, and the rest once per minute as a good starting point for future changes.  That's also why the algorithm is configurable (by passing in boundaries).

When we get specific request on how to configure it, we'll handle that with a config file entry or perhaps a new generator that the config file will reference.  Till then, we need something, and this is a "first guess".

Reply:
Sorry, I missed one question.

> I assume that these events are fairly distributed in practice....

I don't assume that.  Most of these events are reactions to the controller manager adding in tons of entries, modifying them, and removing them in conjunction with kubernetes job launches.  While that doesn't mean that all jobs in kubernetes launch at the same moment, my past experience with building enterprise scale schedulers is that it will be "a series of bursts of entry changes, with each group of near-simultaneous entries distributed over time"

And the hashing distributing over a boundary means that we don't need to add jitter, and we really don't want too much jitter anyway, as it's more valuable to have the database under a constant low load than to see it experience spikes and valleys in tracked events over different periods of time.

Reply:
I'll add in example of different polling algorithms in the comments

- linear: (5, 10, 15, 20, 25, ...)
- exponential (2, 4, 8, 16, 32, 64, ....)

but you can really tune this to be anything, if you just pick the offsets that match. (cosine, sine)

Reply:
The code for boundaries in the polling has been removed.  I'm resolving this conversation.